src/Nethermind/Nethermind.Blockchain/Producers/DevBlockProducer.cs

```diff
@@ -46,25 +47,47 @@ public class DevBlockProducer : BaseBlockProducer
         public override void Start()
         {
             _txPool.NewPending += OnNewPendingTx;
+            BlockTree.NewHeadBlock += OnNewHeadBlock;
         }
-
+        
         public override async Task StopAsync()
         {
             _txPool.NewPending -= OnNewPendingTx;
+            BlockTree.NewHeadBlock -= OnNewHeadBlock;
             await Task.CompletedTask;
         }
 
         protected override UInt256 CalculateDifficulty(BlockHeader parent, UInt256 timestamp) => 1;
 
         private void OnNewPendingTx(object sender, TxEventArgs e)
         {
-            TryProduceNewBlock(CancellationToken.None).ContinueWith(t =>
+            OnNewPendingTxAsync(e);
+        }
+
+        private async void OnNewPendingTxAsync(TxEventArgs e)
+        {
+            _newBlockLock.Wait(TimeSpan.FromSeconds(1));
+            try
             {
-                if (t.IsFaulted)
+                if (!await TryProduceNewBlock(CancellationToken.None))
                 {
-                    if(Logger.IsError) Logger.Error($"Failed to produce block after receiving transaction {e.Transaction}", t.Exception);
+                    _newBlockLock.Release();
                 }
-            });
+            }
+            catch (Exception exception)
+            {
+                if (Logger.IsError) Logger.Error($"Failed to produce block after receiving transaction {e.Transaction}", exception);
+                _newBlockLock.Release();
+            }
+        }
+
+        private void OnNewHeadBlock(object sender, BlockEventArgs e)
+        {
+            if (_newBlockLock.CurrentCount == 0)
```

Comment:
this is not thread safe

Reply:
not sure why are you only releasing when count is 0?

Reply:
MaxCount in semaphore is 1. So we will never go over it. So its fine to release in non-thread safe way.